src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,529 @@
+} RelationAccessHashEntry;
+
+static HTAB *RelationAccessHash;
+
+
+static void RecordPlacementAccessToCache(Oid relationId,
+										 ShardPlacementAccessType accessType);
+static RelationAccessMode GetRelationAccessMode(Oid relationId,
+												ShardPlacementAccessType accessType);
+static void RecordParallelRelationAccess(Oid relationId, ShardPlacementAccessType
+										 placementAccess);
+static void RecordParallelRelationAccessToCache(Oid relationId,
+												ShardPlacementAccessType placementAccess);
+
+
+/*
+ * Empty RelationAccessHash, without destroying the hash table itself.
+ */
+void
+ResetRelationAccessHash()
+{
+	hash_delete_all(RelationAccessHash);
+}
+
+
+/*
+ * Allocate RelationAccessHash.
+ */
+void
+AllocateRelationAccessHash()
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(RelationAccessHashKey);
+	info.entrysize = sizeof(RelationAccessHashEntry);
+	info.hash = tag_hash;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
+
+	RelationAccessHash = hash_create("citus connection cache (relationid)",
+									 8, &info, hashFlags);
+}
+
+
+/*
+ * AssociatePlacementAccessWithRelation associates the placement access to the
+ * distributed relation that the placement belongs to.
+ */
+void
+AssociatePlacementAccessWithRelation(ShardPlacement *placement,
+									 ShardPlacementAccessType accessType)
+{
+	uint64 shardId = INVALID_SHARD_ID;
+	Oid relationId = InvalidOid;
+
+	if (!ShouldRecordRelationAccess())
+	{
+		return;
+	}
+
+	shardId = placement->shardId;
+	relationId = RelationIdForShard(shardId);
+
+	/*
+	 * If a relation is partitioned, record accesses to all of its partitions as well.
+	 * We prefer to use PartitionedTableNoLock() because at this point the necessary
+	 * locks on the relation has already been acquired.
+	 */
+	if (PartitionedTableNoLock(relationId))
+	{
+		List *partitionList = PartitionList(relationId);
+		ShardInterval *shardInterval = LoadShardInterval(shardId);
+		int shardIndex = shardInterval->shardIndex;
+		ListCell *partitionCell = NULL;
+
+		foreach(partitionCell, partitionList)
+		{
+			Oid partitionOid = lfirst_oid(partitionCell);
+			uint64 partitionShardId = INVALID_SHARD_ID;
+			List *partitionShardPlacementList = NIL;
+			ListCell *partitionCell = NULL;
+
+			/*
+			 * During create_distributed_table, the partitions may not
+			 * have been created yet.
+			 */
+			if (!IsDistributedTable(partitionOid))
+			{
+				continue;
+			}
+
+			partitionShardId =
+				ColocatedShardIdInRelation(partitionOid, shardIndex);
+			partitionShardPlacementList = ShardPlacementList(partitionShardId);
+
+			foreach(partitionCell, partitionShardPlacementList)
+			{
+				ShardPlacement *shardPlacement =
+					(ShardPlacement *) lfirst(partitionCell);
+
+				/* recursively record all relation accesses of its partitions */
+				AssociatePlacementAccessWithRelation(shardPlacement, accessType);
```

Comment:
I'm a bit confused as to why we need to do this for all placements, since we ultimately only care about the relationId.

Reply:
That's true, looping on the `placementList` is unnecessary.

But, I preferred to recursively call `AssociatePlacementAccessWithRelation ()` to be able to support multi-level partitioned tables. 

(though we currently don't support those). I thought if we one day decide to implement that, this would probably be forgotten.

I added some comment about that.